Update of subgrid-scale cloud physics for MYNN/GF

climbfuji opened the pull request:
This PR updates the subgrid-scale cloud calculation for MYNN/GF and renames the modules from `mynnrad_{pre,post}` to `sgscloud_rad{pre,post}` for future use by other physics schemes than MYNN. The PR also adds a new variable required for GSDCHEM coupling, qci_conv to the subgrid-scale cloud calculation and the Grell-Freitas deep convection scheme.

Associated PRs:
https://github.com/NOAA-GSD/ccpp-physics/pull/2
https://github.com/NOAA-GSD/fv3atm/pull/3
https://github.com/NOAA-GSD/ufs-weather-model/pull/2

For regression testing, see https://github.com/NOAA-GSD/fv3atm/pull/3.

haiqinli committed:
"to include GF updates in GSDv0beta4"

haiqinli committed:
"update input of imfdeepcnv following Dom's suggestions"

haiqinli committed:
"clean the code"

haiqinli committed:
"update to pass ccpp_gsd regression test"

haiqinli committed:
"update to pass the ccpp_gsd_noah_repro regression test case"

haiqinli committed:
"update with convective cloud & recompute cloud fraction"

climbfuji committed:
Merge branch 'gsd/develop' of https://github.com/noaa-gsd/ccpp-physics into HEAD

climbfuji committed:
Bugfix following merge of gsd/develop into @haiqinli's code

climbfuji commented:
@haiqinli I created a placeholder for the proposed changes (your original PR https://github.com/climbfuji/ccpp-physics/pull/2 update to the head of gsd/develop and pointing to gsd/develop's new home on the NOAA-GSD GitHub page). This PR must be accompanied by changes in fv3atm (at least the suite definition files and the CCPP prebuild config because of the change of the scheme names from MYNNrad_pre to SGSCloud_RadPre, same for post). Does this PR exist somewhere?

haiqinli commented:
@climbfuji I just modified the SDF and prebuild config in my local copy, and have not created this PR. Do you want me to create a PR for this?

climbfuji commented:
> @climbfuji I just modified the SDF and prebuild config in my local copy, and have not created this PR. Do you want me to create a PR for this?

In this case it is ok if you give me access to the files you have changed, and I will create the FV3 PR. Thanks!

haiqinli commented:
@climbfuji I will send you the path of these files on Hera  when Hera is back. Thanks.

haiqinli commented:
@climbfuji The CCPP prebuild config file is /scratch2/BMC/gsd-fv3-dev/Haiqin.Li/gsd-ccpp_heraT/sorc/fv3gfs_ccpp.fd/ccpp/config/ccpp_prebuild_config.py. The SDFs are at /scratch2/BMC/gsd-fv3-dev/Haiqin.Li/gsd-ccpp_heraT/sorc/fv3gfs_ccpp.fd/ccpp/suites. Thanks a lot.

climbfuji commented:
> @climbfuji The CCPP prebuild config file is /scratch2/BMC/gsd-fv3-dev/Haiqin.Li/gsd-ccpp_heraT/sorc/fv3gfs_ccpp.fd/ccpp/config/ccpp_prebuild_config.py. The SDFs are at /scratch2/BMC/gsd-fv3-dev/Haiqin.Li/gsd-ccpp_heraT/sorc/fv3gfs_ccpp.fd/ccpp/suites. Thanks a lot.

Thanks. Can you please let me know which of the suite definition files you changed? All of them are different from what it is gsd/develop today, because your code base is far behind. Thanks!

haiqinli commented:
@climbfuji The followings SDFs are changed. Thanks.
suite_FV3_GFS_v15_gf_mynn.xml:      <scheme>sgscloud_radpre</scheme>
suite_FV3_GFS_v15_mynn.xml:      <scheme>sgscloud_radpre</scheme>
suite_FV3_GFS_v15_thompson_mynn.xml:      <scheme>sgscloud_radpre</scheme>
suite_FV3_GSD_noah.xml:      <scheme>sgscloud_radpre</scheme>
suite_FV3_GSD_v0_nothmp.xml:      <scheme>sgscloud_radpre</scheme>
suite_FV3_GSD_v0_satadj.xml:      <scheme>sgscloud_radpre</scheme>
suite_FV3_GSD_v0.xml:      <scheme>sgscloud_radpre</scheme>

climbfuji commented:
> @climbfuji The followings SDFs are changed. Thanks.
> suite_FV3_GFS_v15_gf_mynn.xml: sgscloud_radpre
> suite_FV3_GFS_v15_mynn.xml: sgscloud_radpre
> suite_FV3_GFS_v15_thompson_mynn.xml: sgscloud_radpre
> suite_FV3_GSD_noah.xml: sgscloud_radpre
> suite_FV3_GSD_v0_nothmp.xml: sgscloud_radpre
> suite_FV3_GSD_v0_satadj.xml: sgscloud_radpre
> suite_FV3_GSD_v0.xml: sgscloud_radpre

Some of these don't exist in the current gsd/develop code. I updated the following SDFs:
```
suite_FV3_GFS_v15_mynn.xml
suite_FV3_GFS_v15_thompson_mynn.xml
suite_FV3_GSD_noah.xml
suite_FV3_GSD_v0.xml
suite_FV3_GSD_SAR.xml
suite_FV3_GSD_v0_drag_suite.xml
```
Basically, I replaced `mynn_rad{pre,post}` with `sgscloud_rad{pre,post}` in all SDFs.

haiqinli commented:
@climbfuji Great! Thanks.

climbfuji commented:
@haiqinli I seem to be missing some changes in your fv3atm repository. ccpp_prebuild.py fails with the error message
```
ERROR: Variable convective_cloud_condesate_after_rainout requested by MODULE_sgscloud_radpre SCHEME_sgscloud_radpre SUBROUTINE_sgscloud_radpre_run & MODULE_cu_gf_driver SCHEME_cu_gf_driver SUBROUTINE_cu_gf_driver_run not provided by the model
```
Can you please provide the missing updates to `GFS_typedefs.*` (and others if you have - I don't think so)? Thanks.

haiqinli commented:
@climbfuji  I am sorry that I forgot to commit this change. Yes, only the update of qci_conv. It is in /scratch2/BMC/gsd-fv3-dev/Haiqin.Li/gsd-ccpp_heraT/sorc/fv3gfs_ccpp.fd/FV3/gfsphysics/GFS_layer/GFS_typedefs.F90 (meta).

climbfuji commented:
Thanks. I will make slight changes (only allocate the array when GF deep convection is used and modify the variable declaration in `sgscloud_radpre` accordingly.

climbfuji committed:
Update of physics/module_SGSCloud_RadPre.{F90,meta}: use imfdeepcnv_gf instead of hardcoded value 3

climbfuji committed:
physics/GFS_rrtmg_pre.F90, physics/radiation_clouds.f: make qci_conv an optional argument for progcld5, because it may not be allocated

DomHeinzeller commented:
@haiqinli @joeolson42 @tanyasmirnova please review this PR and post your comment here (I cannot add you as official reviewers). Thanks.

haiqinli reviewed (approved):
It looks good to me. Thanks.

tanyasmirnova reviewed (commented):
@haiqinli @climbfuji
It looks to me that in the MODULE sgscloud_radpre, the PBL clouds will overwrite the qc,qi that contain convective clouds and computed on lines 113,114. Is it the intent?
Also, I think it could be better first to collect all clouds in QC an QI, and then compute effective radii and optical paths.
Thanks,
Tanya

tanyasmirnova reviewed (commented):
@haiqinli @joeolson42 @climbfuji 
Haiqin and Joe,
Also, if you pass into progcld5 the convective clouds, why don't we pass in there also the PBL sub-grid clouds? I think the cloud-radiation interaction in GSD suite is not correct, and I am testing a different way of coupling: first call the MYNNrad_pre, and after that call GFS_rrtmg_pre, when QC and QI has all clouds in there, and cloud fraction is also computed from MYNN (by the way, do we need to add convective clouds to cloud fraction?). With this sequence of calls, we have to call progcld5 (or progclduni) with cld_uni=.true. to prevent overwriting the cloud fraction. 
Does it make more sense to do it this way?

haiqinli commented:
Tanya,

In progcld5, it computes cloud fraction with cloud condensate (from microphysics & convection) by using Xu & Randall method. The PBL cloud fraction is computed in MYNN. Yes, the BPL cloud fractions overwrite the Xu & Randall cloud fraction unless the qr/qs > 1.e-7.  It is a mixture of clouds.

Then the total/high/mid/low diagnostic clouds is recomputed to include both the Xu&Randall and PBL mixed cloud fractions.

I agree it would be perfect if we can compute the cloud fractions in one place.

Haiqin

DomHeinzeller reviewed (changes requested):
This PR cannot be merged (see https://github.com/NOAA-GSD/fv3atm/pull/3 for details).

haiqinli commented:
@tanyasmirnova @climbfuji Please review Joe's changes at /scratch2/BMC/gsd-fv3-dev/Haiqin.Li/NEMSfv3gfs/gsd_develop_hli/ccpp/physics/physics/module_SGSCloud_RadPre.F90 (meta). We also submitted a set of retro runs, but they are still waiting in the queue. Will update the evaluation when the runs are done. Thanks.

tanyasmirnova commented:
Haiqin,
There is a test case that Dom and I use to test the code changes. Should we test the changes for this test case before running the full retro?

haiqinli commented:
Hi Tanaya,

Please update the
/scratch2/BMC/gsd-fv3-dev/Haiqin.Li/NEMSfv3gfs/gsd_develop_hli/ccpp/physics/physics/module_SGSCloud_RadPre.F90,
which
Joe just did an update. The module_MYNNRad_pre.F90, module_MYNNRad_post.F90
are replaced by module_SGSCloud_RadPre.F90 and module_SGSCloud_RadPre.F90.

Best Wishes
Haiqin

On Tue, Dec 17, 2019 at 10:57 AM tanyasmirnova <notifications@github.com>
wrote:

> Haiqin,
> There is a test case that Dom and I use to test the code changes. Should
> we test the changes for this test case before running the full retro?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GSD/ccpp-physics/pull/2?email_source=notifications&email_token=AJHAAOAVD2GTCN6JWG3QXBLQZEHIHA5CNFSM4J2EJV7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHDM2SQ#issuecomment-566676810>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJHAAOERXIJWQ65UZ7CSR7TQZEHIHANCNFSM4J2EJV7A>
> .
>

climbfuji commented:
@haiqinli @tanyasmirnova I got segmentation faults for some of the tests, and it seems that this has got to do with the "empty" if block:
```
         IF (qr(i,k) > 1.0e-7 .OR. qs(i,k) > 1.0e-7 .OR. qci_conv(i,k)>1.0e-7) THEN
             !Keep Xu-RandalL clouds fraction - do not overwrite
         ELSE
             clouds1(i,k) = CLDFRA_BL(i,k)
         ENDIF
```
I changed this logic as shown below, now rerunning the tests:
```
         IF (.not.(qr(i,k) > 1.0e-7 .OR. qs(i,k) > 1.0e-7 .OR. qci_conv(i,k)>1.0e-7)) THEN
             clouds1(i,k) = CLDFRA_BL(i,k)
         !ELSE
             !Keep Xu-RandalL clouds fraction - do not overwrite
         ENDIF
```
I thought that empty execution blocks were ok, but maybe I was wrong - will let you know.

climbfuji committed:
physics/module_SGSCloud_RadPre.*: add logic for qci_conv when Grell-Freitas convection scheme is used

climbfuji commented on physics/module_SGSCloud_RadPre.F90:
```diff
@@ -131,12 +132,13 @@ SUBROUTINE sgscloud_radpre_run(            &
         do i = 1, im
 
-         IF( qr(i,k) > 1.0e-7 .OR. qs(i,k) > 1.0e-7)THEN
+         IF (qr(i,k) > 1.0e-7 .OR. qs(i,k) > 1.0e-7 .OR. (imfdeepcnv == imfdeepcnv_gf .AND. qci_conv(i,k)>1.0e-7) THEN
```
@haiqinli please check that the logic I added here makes sense. Without testing for imfdeepcnv==imfdeepcnv_gf first, the code crashes when GF is not used (because qci_conv is not allocated).

haiqinli commented:
@climbfuji Yes, you are right. I only tested it with GSD suite, which includes both MYNN and GF.

climbfuji commented:
At last! Regression testing on hera/intel using rt_ccpp_gsd.conf after all changes succeeded (first create, then verify): all tests pass, including the warmstart test.

[rt_ccpp_gsd_create.log](https://github.com/NOAA-GSD/ccpp-physics/files/3986143/rt_ccpp_gsd_create.log)
[rt_ccpp_gsd_verify.log](https://github.com/NOAA-GSD/ccpp-physics/files/3986144/rt_ccpp_gsd_verify.log)

climbfuji commented:
@tanyasmirnova @haiqinli @joeolson42 can you please confirm and approve that this version is good to merge? Thank you!

DomHeinzeller commented:
Superseded by #12
